src/state/routing/useRoutes.ts

```diff
@@ -0,0 +1,74 @@
+import { Currency, CurrencyAmount, Ether, Token } from '@uniswap/sdk-core'
+import { FeeAmount, Pool, Route } from '@uniswap/v3-sdk'
+import { useMemo } from 'react'
+import { GetQuoteResult } from 'state/routing/slice'
+
+export function useRoutes(
+  currencyIn: Currency | undefined,
+  currencyOut: Currency | undefined,
+  quoteResult: Pick<GetQuoteResult, 'route'> | undefined
+):
+  | {
+      route: Route<Currency, Currency>
+      inputAmount: CurrencyAmount<Currency>
+      outputAmount: CurrencyAmount<Currency>
+    }[]
+  | undefined {
+  return useMemo(() => {
+    if (!quoteResult || !quoteResult.route || !currencyIn || !currencyOut) return undefined
+
+    if (quoteResult.route.length === 0) return []
+
+    const parsedCurrencyIn = currencyIn.isNative
+      ? Ether.onChain(currencyIn.chainId)
+      : parseToken(quoteResult.route[0][0].tokenIn)
+
+    const parsedCurrencyOut = currencyOut.isNative
+      ? Ether.onChain(currencyOut.chainId)
+      : parseToken(quoteResult.route[0][quoteResult.route[0].length - 1].tokenOut)
+
+    try {
+      return quoteResult.route.map((route) => {
+        const rawAmountIn = route[0].amountIn
+        const rawAmountOut = route[route.length - 1].amountOut
+
+        if (!rawAmountIn || !rawAmountOut) {
+          throw new Error('Expected both amountIn and amountOut to be present')
+        }
+
+        return {
+          route: new Route(route.map(parsePool), parsedCurrencyIn, parsedCurrencyOut),
+          inputAmount: CurrencyAmount.fromRawAmount(parsedCurrencyIn, rawAmountIn),
+          outputAmount: CurrencyAmount.fromRawAmount(parsedCurrencyOut, rawAmountOut),
+        }
+      })
+    } catch (e) {
+      // `Route` constructor may throw if inputs/outputs are temporarily out of sync
+      // (RTK-Query always returns the latest data which may not be the right inputs/outputs)
+      // This is not fatal and will fix itself in future render cycles
+      console.debug(e)
+      return undefined
+    }
+  }, [currencyIn, currencyOut, quoteResult])
+}
+
+const parseToken = ({ address, chainId, decimals, symbol }: GetQuoteResult['route'][0][0]['tokenIn']): Currency => {
```

Comment:
should probably be `[number]` instead of `[0]`?

Reply:
why doesn't this just return Token instead of Currency and you don't need wrapped below

Reply:
`Currency` is helpful because the input token may be native.

I don't think that `GetQuoteResult['quote'][number]` works? Is that what you mean?

Reply:
but doesn't the parseToken method always return a Token? i don't see how it can return a currency with this implementation

Reply:
Ah right, changed to `Token`.